cdi-controller in crashing because certain type of ingress objects

<!-- This form is for bug reports and feature requests ONLY!
Also make sure that you visit our User Guide at https://kubevirt.io/user-guide/
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:
 /kind bug


**What happened**:
cdi-controller fails with following error
```
 E0722 23:32:40.307785       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1543 [running]:
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x171f060, 0x29b6db0)
  vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0x92
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
  vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x171f060, 0x29b6db0)
  GOROOT/src/runtime/panic.go:969 +0x166
kubevirt.io/containerized-data-importer/pkg/controller.getURLFromIngress(0xc003ee4310, 0xc000044044, 0xf, 0x0, 0x0)
  pkg/controller/config-controller.go:600 +0x204
kubevirt.io/containerized-data-importer/pkg/controller.watchIngress.func2(0x1bc3780, 0xc003ee4310, 0x1bc3780)
  pkg/controller/config-controller.go:533 +0x64
kubevirt.io/containerized-data-importer/vendor/sigs.k8s.io/controller-runtime/pkg/predicate.Funcs.Create(...)
  vendor/sigs.k8s.io/controller-runtime/pkg/predicate/predicate.go:71
kubevirt.io/containerized-data-importer/vendor/sigs.k8s.io/controller-runtime/pkg/source/internal.EventHandler.OnAdd(0x1b92f80, 0xc000542a58, 0x1baadc0, 0xc000a7a480, 0xc00099bf90, 0x1, 0x1, 0x18e76e0, 0xc003ee4310)
  vendor/sigs.k8s.io/controller-runtime/pkg/source/internal/eventsource.go:57 +0xd1
kubevirt.io/containerized-data-importer/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
  vendor/k8s.io/client-go/tools/cache/shared_informer.go:777 +0x129
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000617760)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00033ef60, 0x1b55bc0, 0xc002d92000, 0x16e8801, 0xc002d5c000)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xa3
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000617760, 0x3b9aca00, 0x0, 0xc002d4c001, 0xc002d5c000)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
kubevirt.io/containerized-data-importer/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc000860180)
  vendor/k8s.io/client-go/tools/cache/shared_informer.go:771 +0x95
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0002f27d0, 0xc0027e2ae0)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x51
created by kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x62
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
  panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1583084]

goroutine 1543 [running]:
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
  vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x171f060, 0x29b6db0)
  GOROOT/src/runtime/panic.go:969 +0x166
kubevirt.io/containerized-data-importer/pkg/controller.getURLFromIngress(0xc003ee4310, 0xc000044044, 0xf, 0x0, 0x0)
  pkg/controller/config-controller.go:600 +0x204
kubevirt.io/containerized-data-importer/pkg/controller.watchIngress.func2(0x1bc3780, 0xc003ee4310, 0x1bc3780)
  pkg/controller/config-controller.go:533 +0x64
kubevirt.io/containerized-data-importer/vendor/sigs.k8s.io/controller-runtime/pkg/predicate.Funcs.Create(...)
  vendor/sigs.k8s.io/controller-runtime/pkg/predicate/predicate.go:71
kubevirt.io/containerized-data-importer/vendor/sigs.k8s.io/controller-runtime/pkg/source/internal.EventHandler.OnAdd(0x1b92f80, 0xc000542a58, 0x1baadc0, 0xc000a7a480, 0xc00099bf90, 0x1, 0x1, 0x18e76e0, 0xc003ee4310)
  vendor/sigs.k8s.io/controller-runtime/pkg/source/internal/eventsource.go:57 +0xd1
kubevirt.io/containerized-data-importer/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
  vendor/k8s.io/client-go/tools/cache/shared_informer.go:777 +0x129
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000617760)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000d8df60, 0x1b55bc0, 0xc002d92000, 0x16e8801, 0xc002d5c000)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xa3
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000617760, 0x3b9aca00, 0x0, 0xc002d4c001, 0xc002d5c000)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
kubevirt.io/containerized-data-importer/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc000860180)
  vendor/k8s.io/client-go/tools/cache/shared_informer.go:771 +0x95
kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0002f27d0, 0xc0027e2ae0)
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x51
created by kubevirt.io/containerized-data-importer/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
  vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x62
```
When there is an ingress object as below on the cluster which has `Spec.DefaultBackend.Service` as empty then we hit the error
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: test-ingress
  annotations:
    nginx.org/ssl-services: "cdi-uploadproxy"
    ingress.kubernetes.io/ssl-passthrough: "true"
    nginx.ingress.kubernetes.io/secure-backends: "true"
    nginx.ingress.kubernetes.io/proxy-body-size: "0"
spec:
  defaultBackend:
    resource:
      apiGroup: test.test
      kind: test
      name: test
  tls:
  - hosts:
    # change to a valid FQDN in your organization
    - cdi-uploadproxy.example.com
```
It happening because [this](https://github.com/kubevirt/containerized-data-importer/blob/466f706bf190a03431653107a430ab509b5b62f9/pkg/controller/config-controller.go#L600) check.

I am not sure why we need to iterate through all the ingress objects on cluster
**What you expected to happen**:
cdi-controller should run without error

**How to reproduce it (as minimally and precisely as possible)**:
Run cdi with following ingress object on cluster
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: test-ingress
  annotations:
    nginx.org/ssl-services: "cdi-uploadproxy"
    ingress.kubernetes.io/ssl-passthrough: "true"
    nginx.ingress.kubernetes.io/secure-backends: "true"
    nginx.ingress.kubernetes.io/proxy-body-size: "0"
spec:
  defaultBackend:
    resource:
      apiGroup: test.test
      kind: test
      name: test
  tls:
  - hosts:
    # change to a valid FQDN in your organization
    - cdi-uploadproxy.example.com
```
**Anything else we need to know?**:

**Environment**:
- CDI version (use `kubectl get deployments cdi-deployment -o yaml`):
v1.36.0
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Install tools:
- Others:

--- a/pkg/controller/config-controller.go
+++ b/pkg/controller/config-controller.go
@@ -596,7 +596,7 @@ func watchClusterProxy(mgr manager.Manager, configController controller.Controll
 }
 
 func getURLFromIngress(ing *networkingv1.Ingress, uploadProxyServiceName string) string {
-	if ing.Spec.DefaultBackend != nil {
+	if ing.Spec.DefaultBackend != nil && ing.Spec.DefaultBackend.Service != nil {
 		if ing.Spec.DefaultBackend.Service.Name != uploadProxyServiceName {
 			return ""
 		}
@@ -607,7 +607,7 @@ func getURLFromIngress(ing *networkingv1.Ingress, uploadProxyServiceName string)
 			continue
 		}
 		for _, path := range rule.HTTP.Paths {
-			if path.Backend.Service.Name == uploadProxyServiceName {
+			if path.Backend.Service != nil && path.Backend.Service.Name == uploadProxyServiceName {
 				if rule.Host != "" {
 					return rule.Host
 				}
@@ -615,7 +615,6 @@ func getURLFromIngress(ing *networkingv1.Ingress, uploadProxyServiceName string)
 		}
 	}
 	return ""
-
 }
 
 func getURLFromRoute(route *routev1.Route, uploadProxyServiceName string) string {

--- a/pkg/controller/config-controller_test.go
+++ b/pkg/controller/config-controller_test.go
@@ -172,6 +172,22 @@ var _ = Describe("Controller ingress reconcile loop", func() {
 		Expect(err).ToNot(HaveOccurred())
 		Expect(*cdiConfig.Status.UploadProxyURL).To(Equal(testURL))
 	})
+
+	DescribeTable("Should not set proxyURL if invalid ingress exists", func(createIngress func(name, ns, service, url string) *networkingv1.Ingress) {
+		reconciler, cdiConfig := createConfigReconciler(createIngressList(
+			*createIngress("test-ingress", "test-ns", "service", testURL),
+		))
+		reconciler.uploadProxyServiceName = testServiceName
+		err := reconciler.reconcileIngress(cdiConfig)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(cdiConfig).ToNot(BeNil())
+		Expect(cdiConfig.Status).ToNot(BeNil())
+		Expect(cdiConfig.Status.UploadProxyURL).To(BeNil())
+	},
+		Entry("No default backend", createNoDefaultBackendIngress),
+		Entry("No service", createNoServiceIngress),
+		Entry("0 rules", createNoRulesIngress),
+	)
 })
 
 var _ = Describe("Controller route reconcile loop", func() {
@@ -837,6 +853,24 @@ func createIngress(name, ns, service, url string) *networkingv1.Ingress {
 			},
 		},
 	}
+}
+
+func createNoDefaultBackendIngress(name, ns, service, url string) *networkingv1.Ingress {
+	res := createIngress(name, ns, service, url)
+	res.Spec.DefaultBackend = nil
+	return res
+}
+
+func createNoServiceIngress(name, ns, service, url string) *networkingv1.Ingress {
+	res := createIngress(name, ns, service, url)
+	res.Spec.DefaultBackend.Service = nil
+	return res
+}
+
+func createNoRulesIngress(name, ns, service, url string) *networkingv1.Ingress {
+	res := createIngress(name, ns, service, url)
+	res.Spec.Rules = []networkingv1.IngressRule{}
+	return res
 }
 
 func createConfigMap(name, namespace string) *corev1.ConfigMap {

--- a/tests/cdiconfig_test.go
+++ b/tests/cdiconfig_test.go
@@ -13,6 +13,7 @@ import (
 
 	"github.com/coreos/go-semver/semver"
 	route1client "github.com/openshift/client-go/route/clientset/versioned"
+	v1 "k8s.io/api/core/v1"
 	networkingv1 "k8s.io/api/networking/v1"
 	storagev1 "k8s.io/api/storage/v1"
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
@@ -414,6 +415,19 @@ var _ = Describe("CDI ingress config tests", func() {
 			return *config.Status.UploadProxyURL
 		}, time.Second*30, time.Second).Should(Equal(override))
 	})
+
+	It("Should not crash and uploadProxy empty, if ingress defined without service", func() {
+		ingress = createIngress("test-ingress", f.CdiInstallNs, "cdi-uploadproxy", ingressUrl)
+		// The cdi controller will not process this ingress because it doesn't have the url we are looking for.
+		_, err := f.K8sClient.NetworkingV1().Ingresses(f.CdiInstallNs).Create(context.TODO(), ingress, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		By("Expecting uploadproxy url to be blank")
+		Eventually(func() bool {
+			config, err := f.CdiClient.CdiV1beta1().CDIConfigs().Get(context.TODO(), common.ConfigName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			return config.Status.UploadProxyURL == nil
+		}, time.Second*30, time.Second).Should(BeTrue())
+	})
 })
 
 var _ = Describe("CDI route config tests", func() {
@@ -601,6 +615,41 @@ func createIngress(name, ns, service, hostUrl string) *networkingv1.Ingress {
 			},
 			Rules: []networkingv1.IngressRule{
 				{Host: hostUrl},
+			},
+		},
+	}
+}
+
+func createNoServiceIngress(name, ns string) *networkingv1.Ingress {
+	apiGroup := "test.test"
+	return &networkingv1.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "networking/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+			Labels: map[string]string{
+				"nginx.org/ssl-services":                      "cdi-uploadproxy",
+				"ingress.kubernetes.io/ssl-passthrough":       "true",
+				"nginx.ingress.kubernetes.io/secure-backends": "true",
+				"nginx.ingress.kubernetes.io/proxy-body-size": "0",
+			},
+		},
+		Spec: networkingv1.IngressSpec{
+			DefaultBackend: &networkingv1.IngressBackend{
+				Resource: &v1.TypedLocalObjectReference{
+					APIGroup: &apiGroup,
+					Kind:     "test",
+					Name:     "test",
+				},
+			},
+			TLS: []networkingv1.IngressTLS{
+				{
+					Hosts: []string{
+						"cdi-uploadproxy.example.com",
+					},
+				},
 			},
 		},
 	}

